src/aiida_quantumespresso/workflows/ph/base.py

```diff
@@ -257,13 +258,39 @@ def handle_out_of_walltime(self, node):
     @process_handler(priority=410, exit_codes=PhCalculation.exit_codes.ERROR_CONVERGENCE_NOT_REACHED)
     def handle_convergence_not_reached(self, node):
         """Handle `ERROR_CONVERGENCE_NOT_REACHED` exit code: decrease the mixing beta and restart."""
+        self.ctx.restart_calc = node
         factor = self.defaults.delta_factor_alpha_mix
+
+        nmix_ph = self.ctx.inputs.parameters.get('INPUTPH', {}).get('nmix_ph', self.defaults.nmix_ph)
+
+        if nmix_ph < 20:  # in all these cases, better trying to use more iterations in the mixing first
+            if nmix_ph < 8:  # 8~20 is the recommended range on the Quantum ESPRESSO documentation
+                nmix_ph_new = 8
+            if nmix_ph >= 8:
+                nmix_ph_new = 20
+
+            self.ctx.inputs.parameters.setdefault('INPUTPH', {})['nmix_ph'] = nmix_ph_new
+
+            action = f'increased number of mixing iteration from {nmix_ph} to {nmix_ph_new} and restarting'
+            self.report_error_handled(node, action)
+
+            return ProcessHandlerReport(True)
+
+        # now try playing with the mixing parameters
         alpha_mix = self.ctx.inputs.parameters.get('INPUTPH', {}).get('alpha_mix(1)', self.defaults.alpha_mix)
         alpha_mix_new = alpha_mix * factor
 
-        self.ctx.restart_calc = node
-        self.ctx.inputs.parameters.setdefault('INPUTPH', {})['alpha_mix(1)'] = alpha_mix_new
+        if self.ctx.inputs.parameters.get('INPUTPH', {}).get('alpha_mix(20)', -1) != -1:
+            action = 'no more efficient strategies for convergence.'
+            self.report_error_handled(node, action)
+            return ProcessHandlerReport(False)
+
+        if alpha_mix_new >= 0.1:
+            self.ctx.inputs.parameters.setdefault('INPUTPH', {})['alpha_mix(1)'] = alpha_mix_new
+            action = f'reduced alpha_mix from {alpha_mix} to {alpha_mix_new} and restarting'
+        else:
+            self.ctx.inputs.parameters.setdefault('INPUTPH', {})['alpha_mix(20)'] = 0.4
```

Comment:
Setting `alpha_mix(20)` to some value means that _from iteration 20_ the mixing factor will be adapted, right?

If in this case `alpha_mix(1)` is already e.g 0.1, it would help to suddenly increase the mixing at iteration 20?

Reply:
Yes, exactly. I remember seeing this being suggested in the QE forum for some tricky cases. The idea is to do have low mixing at the beginning, and then at the 20-th iteration one expect to have the mixed solution quite sabilized and then one can mix more the self-consistent solution (IDK if I explained myself).

Reply:
Yes, I see how this makes sense. Curious to see it in action. ^^

I suppose we should gather some statistics similar as in https://github.com/aiidateam/aiida-quantumespresso/issues/961. I have a good number of `ph.x` runs, but would still have to parse the convergence info. Let's adapt the strategy for now and maybe revisit!